### ISSUE TITLE
autowrap cython fail on matrix

### ISSUE THREAD

**Issue by edwintye** [sympy/sympy#9098]
I get an error when trying to autowrap a matrix with cython, code as follows

```python
from sympy import sin,Symbol,zeros
from sympy.utilities.autowrap import ufuncify,autowrap

x = Symbol('x')
expr = sin(x)/x

a = zeros(2,1)
for i in range(0,2): a[i]  = expr

aF2py = autowrap(a,args=[x])
aCython = autowrap(a,args=[x],backend='Cython',verbose=True)
```

with error:

<ipython-input-6-c484d73f5fe4> in <module>()
----> 1 aCython = autowrap(a,args=[x],backend='Cython',verbose=True)

/home_nfs/edwint/.local/lib/python2.7/site-packages/sympy/core/cache.pyc in wrapper(*args, **kwargs)
     89             def wrapper(*args, **kwargs):
     90                 try:
---> 91                     retval = cfunc(*args, **kwargs)
     92                 except TypeError:
     93                     retval = func(*args, **kwargs)

/home_nfs/edwint/.local/lib/python2.7/site-packages/sympy/core/compatibility.pyc in wrapper(*args, **kwds)
    855                 except TypeError:
    856                     stats[MISSES] += 1
--> 857                     return user_function(*args, **kwds)
    858                 with lock:
    859                     link = cache_get(key)

/home_nfs/edwint/.local/lib/python2.7/site-packages/sympy/utilities/autowrap.pyc in autowrap(expr, language, backend, tempdir, args, flags, verbose, helpers)
    506         helps.append(make_routine(name, expr, args))
    507 
--> 508     return code_wrapper.wrap_code(routine, helpers=helps)
    509 
    510 

/home_nfs/edwint/.local/lib/python2.7/site-packages/sympy/utilities/autowrap.pyc in wrap_code(self, routine, helpers)
    142             self._generate_code(routine, helpers)
    143             self._prepare_files(routine)
--> 144             self._process_files(routine)
    145             mod = __import__(self.module_name)
    146         finally:

/home_nfs/edwint/.local/lib/python2.7/site-packages/sympy/utilities/autowrap.pyc in _process_files(self, routine)
    161             raise CodeWrapError(
    162                 "Error while executing command: %s. Command output is:\n%s" % (
--> 163                     " ".join(command), e.output.decode()))
    164         if not self.quiet:
    165             print(retoutput)

UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 585: ordinal not in range(128)

so I am not sure where it went wrong, and compiling with the f2py backend worked for me.

**Comment by moorepants** [comment]
This runs for me with SymPy 0.7.6 and Cython 0.21.2.

What versions are you running? Also, what OS do you have?

**Comment by edwintye** [comment]
cython 0.22 and sympy 0.7.6.  os is bullx 6.2.  I will try it at home later as well later tonight

**Comment by moorepants** [comment]
I'm guessing bullx is Linux based? I'm running this on Linux too.

Are you using Python 2 or 3?

**Comment by edwintye** [comment]
yes, linux, redhat derivative.  Python 2.7.5

**Comment by moorepants** [comment]
I used Python 2.7.9. I thought the unicode error may be a Python 2/3 thing.

**Comment by moorepants** [comment]
If you run this through the debugger, it would be interesting to know what the command variable is. It seems like a command is being generated that doesn't have all ascii characters.

**Comment by moorepants** [comment]
Actually this command `e.output.decode()` is the error. Finding out what `e.output` is would help.

**Comment by edwintye** [comment]
If I have done it correctly, then e.output is 

running build_ext
skipping 'wrapper_module_21.c' Cython extension (up-to-date)
building 'wrapper_module_21' extension
gcc -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/usr/local/packages/Python/2.7.5/include/python2.7 -c wrapper_module_21.c -o build/temp.linux-x86_64-2.7/wrapper_module_21.o -std=c99
wrapper_module_21.c:250:31: error: numpy/arrayobject.h: No such file or directory
wrapper_module_21.c:251:31: error: numpy/ufuncobject.h: No such file or directory
wrapper_module_21.c:503: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_int8_t’
wrapper_module_21.c:512: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_int16_t’
wrapper_module_21.c:521: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_int32_t’
wrapper_module_21.c:530: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_int64_t’
wrapper_module_21.c:539: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_uint8_t’
wrapper_module_21.c:548: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_uint16_t’
wrapper_module_21.c:557: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_uint32_t’
wrapper_module_21.c:566: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_uint64_t’
wrapper_module_21.c:575: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_float32_t’
wrapper_module_21.c:584: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_float64_t’
wrapper_module_21.c:593: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_int_t’
wrapper_module_21.c:602: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_long_t’
wrapper_module_21.c:611: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_longlong_t’
wrapper_module_21.c:620: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_uint_t’
wrapper_module_21.c:629: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_ulong_t’
wrapper_module_21.c:638: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_ulonglong_t’
wrapper_module_21.c:647: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_intp_t’
wrapper_module_21.c:656: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_uintp_t’
wrapper_module_21.c:665: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_float_t’
wrapper_module_21.c:674: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_double_t’
wrapper_module_21.c:683: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_longdouble_t’
wrapper_module_21.c:714: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_cfloat_t’
wrapper_module_21.c:723: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_cdouble_t’
wrapper_module_21.c:732: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_clongdouble_t’
wrapper_module_21.c:741: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘__pyx_t_5numpy_complex_t’
wrapper_module_21.c:1060: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c:1063: error: ‘__pyx_t_5numpy_double_t’ undeclared here (not in a function)
wrapper_module_21.c:1072: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c:1073: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c: In function ‘__pyx_pf_17wrapper_module_21_autofunc_c’:
wrapper_module_21.c:1180: error: ‘PyArrayObject’ undeclared (first use in this function)
wrapper_module_21.c:1180: error: (Each undeclared identifier is reported only once
wrapper_module_21.c:1180: error: for each function it appears in.)
wrapper_module_21.c:1180: error: ‘__pyx_v_out_5117023170673287343’ undeclared (first use in this function)
wrapper_module_21.c:1187: error: ‘__pyx_t_3’ undeclared (first use in this function)
wrapper_module_21.c:1213: error: expected expression before ‘)’ token
wrapper_module_21.c:1217: error: expected expression before ‘)’ token
wrapper_module_21.c:1223: error: expected expression before ‘)’ token
wrapper_module_21.c: In function ‘__pyx_pw_5numpy_7ndarray_1__getbuffer__’:
wrapper_module_21.c:1286: warning: implicit declaration of function ‘__pyx_pf_5numpy_7ndarray___getbuffer__’
wrapper_module_21.c:1286: error: ‘PyArrayObject’ undeclared (first use in this function)
wrapper_module_21.c:1286: error: expected expression before ‘)’ token
wrapper_module_21.c: At top level:
wrapper_module_21.c:1293: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c: In function ‘__pyx_pw_5numpy_7ndarray_3__releasebuffer__’:
wrapper_module_21.c:2088: warning: implicit declaration of function ‘__pyx_pf_5numpy_7ndarray_2__releasebuffer__’
wrapper_module_21.c:2088: error: ‘PyArrayObject’ undeclared (first use in this function)
wrapper_module_21.c:2088: error: expected expression before ‘)’ token
wrapper_module_21.c: At top level:
wrapper_module_21.c:2094: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c: In function ‘__pyx_f_5numpy_PyArray_MultiIterNew1’:
wrapper_module_21.c:2180: warning: implicit declaration of function ‘PyArray_MultiIterNew’
wrapper_module_21.c:2180: warning: assignment makes pointer from integer without a cast
wrapper_module_21.c: In function ‘__pyx_f_5numpy_PyArray_MultiIterNew2’:
wrapper_module_21.c:2230: warning: assignment makes pointer from integer without a cast
wrapper_module_21.c: In function ‘__pyx_f_5numpy_PyArray_MultiIterNew3’:
wrapper_module_21.c:2280: warning: assignment makes pointer from integer without a cast
wrapper_module_21.c: In function ‘__pyx_f_5numpy_PyArray_MultiIterNew4’:
wrapper_module_21.c:2330: warning: assignment makes pointer from integer without a cast
wrapper_module_21.c: In function ‘__pyx_f_5numpy_PyArray_MultiIterNew5’:
wrapper_module_21.c:2380: warning: assignment makes pointer from integer without a cast
wrapper_module_21.c: At top level:
wrapper_module_21.c:2413: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c:3130: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c:3218: error: expected ‘)’ before ‘*’ token
wrapper_module_21.c: In function ‘initwrapper_module_21’:
wrapper_module_21.c:3527: error: ‘PyArray_Descr’ undeclared (first use in this function)
wrapper_module_21.c:3528: error: ‘PyArrayIterObject’ undeclared (first use in this function)
wrapper_module_21.c:3529: error: ‘PyArrayMultiIterObject’ undeclared (first use in this function)
wrapper_module_21.c:3530: error: ‘PyArrayObject’ undeclared (first use in this function)
wrapper_module_21.c:3531: error: ‘PyUFuncObject’ undeclared (first use in this function)
error: command 'gcc' failed with exit status 1

I have also tried it at home with Fedora and it also failed (may not be the same error).

**Comment by moorepants** [comment]
Do you have numpy installed? Also, the Python headers are required too.

**Comment by moorepants** [comment]
Have you tried the development version yet? There was this bug fix that could be related: https://github.com/sympy/sympy/pull/8848

**Comment by edwintye** [comment]
Yes.  It worked on the development version.  Many thanks.  

On a side note, is there any reason why the numpy compile using lambdify worked?  I presume it also required the numpy header files there?

**Comment by moorepants** [comment]
Lambdify doesn't compile any C level code so the numpy development headers are not necessary. Lambdify also does not require numpy to work, but if numpy is present it will utilize it.

Glad it works now.

Fixed by #8848.
